x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsdbIndexModeSettingsProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.logsdb;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.MetadataIndexTemplateService;
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexSettingProvider;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.cluster.settings.ClusterSettings.CLUSTER_LOGSDB_ENABLED;
+
+final class LogsdbIndexModeSettingsProvider implements IndexSettingProvider {
+    private static final Logger logger = LogManager.getLogger(LogsdbIndexModeSettingsProvider.class);
+    private static final String LOGS_PATTERN = "logs-*-*";
+    private volatile boolean isLogsdbEnabled;
+
+    LogsdbIndexModeSettingsProvider(final Settings settings) {
+        this.isLogsdbEnabled = CLUSTER_LOGSDB_ENABLED.get(settings);
+    }
+
+    public void updateClusterIndexModeLogsdbEnabled(boolean isLogsdbEnabled) {
+        logger.debug("LogsDB " + (isLogsdbEnabled ? "enabled" : "disabled"));
+        this.isLogsdbEnabled = isLogsdbEnabled;
+    }
+
+    @Override
+    public Settings getAdditionalIndexSettings(
+        final String indexName,
+        final String dataStreamName,
+        boolean isTimeSeries,
+        final Metadata metadata,
+        final Instant resolvedAt,
+        final Settings settings,
+        final List<CompressedXContent> combinedTemplateMappings
+    ) {
+        if (isLogsdbEnabled == false || dataStreamName == null) {
+            return Settings.EMPTY;
+        }
+
+        final IndexMode indexMode = resolveIndexMode(settings.get("index.mode"));
+        if (indexMode != null) {
+            return Settings.EMPTY;
+        }
+
+        if (usesLogsAtSettingsComponentTemplate(metadata, dataStreamName) && matchesLogsPattern(dataStreamName)) {
+            return Settings.builder().put("index.mode", IndexMode.LOGSDB.getName()).build();
+        }
+
+        return Settings.EMPTY;
+    }
+
+    private static boolean matchesLogsPattern(final String name) {
+        return Regex.simpleMatch(LOGS_PATTERN, name);
+    }
+
+    private IndexMode resolveIndexMode(final String mode) {
+        return mode == null
```

Comment:
I don't think there is a need here to use streams api. Maybe just do:

```
 IndexMode indexMode = mode != null ? Enum.valueOf(IndexMode.class, mode(Locale.ROOT)) : null;
```

Since we read `mode` variable from settings, it can only have valid values.